Apollo Server 4 Upgrade

nnoce14 opened the pull request:
Made changes to MongoDataSource to be compliant with how data sources were changed in Apollo Server 4. See latest comment on Issue #114 for more information.

devileye98 commented on index.d.ts:
```diff
@@ -1,21 +1,21 @@
-declare module 'apollo-datasource-mongodb' {
-  import { DataSource } from 'apollo-datasource'
+declare module 'apollo-mongo-datasource' {
```
Isnt this supposed to be apollo-datasource-mongodb itself since we are making changes to the same package?

nnoce14 replied:
You're right, I was originally going to create a separate package because I wasn't sure this repo was still being maintained. Just fixed the module name, should be good now

devileye98 replied:
IMO, I think it would be better to maintain this as a separate package, since I dont see this getting merged anytime soon.
This is something my team has also been dependent on.

lorensr commented:
@nnoce14 thanks so much for submitting! and @devileye98 for reviewing 🙏

Do we need any more automated tests to make sure this works with AS v4?

Could you also add a note to the readme saying which versions of this package support which AS versions, and a link to the old (current) version of the README/docs?

devileye98 commented:
> ed tests to make sure this w

I forked out this PR and tested it against our application. Apart from a few minor changes in our types, we were able to get the server up and running and were also able to run our E2E and unit tests

nnoce14 commented on src/__tests__/datasource.test.js:
```diff
@@ -135,4 +132,18 @@ describe('Mongoose', () => {
     expect(res2[0]).toBeUndefined()
   })
+
+  test('Data Source with Context', async () => {
+    const users = new Users({ modelOrCollection: UserModel, context: { token: '123' }})
+    
+    expect(users.context.token).toBe('123')
+  })
+
+  test('Data Source with Context that contains a User', async () => {
+    const users = new Users({ modelOrCollection: userCollection, context: { user: alice }})
+    const user = await users.findOneById(alice._id)
+
+    expect(user.name).toBe('Alice')
+    expect(users.context.user.name).toBe(user.name)
+  })
```
@lorensr I added these two tests to check the behavior for providing Context to datasources in Apollo Server 4. I also made minor changes to the construction of the Users datasource. Let me know if you think more tests are necessary. I also added the old README file and added a note to both which explains the versioning.

devileye98 reviewed (changes requested):
I see that both mongodb and mongoose are on older versions. If we could update it to the newest packages, it would be awesome as we can make use of the newer types from both these packages and get rid of any type errors for users consuming this package.

nnoce14 commented:
> I see that both mongodb and mongoose are on older versions. If we could update it to the newest packages, it would be awesome as we can make use of the newer types from both these packages and get rid of any type errors for users consuming this package.

I just updated mongoose from v5 to v7, which is latest. I also updated mongodb from v3 to v4. However, the latest version is v5 but I was unable to get that to work because of an ongoing issue with BSON v5 found [here](https://github.com/nextauthjs/next-auth/issues/7022)

devileye98 commented:
> > I see that both mongodb and mongoose are on older versions. If we could update it to the newest packages, it would be awesome as we can make use of the newer types from both these packages and get rid of any type errors for users consuming this package.
> 
> I just updated mongoose from v5 to v7, which is latest. I also updated mongodb from v3 to v4. However, the latest version is v5 but I was unable to get that to work because of an ongoing issue with BSON v5 found [here](https://github.com/nextauthjs/next-auth/issues/7022)

I think we resolved this by updating the Bson package. I think that should resolve the issue.

nnoce14 commented:
> > > I see that both mongodb and mongoose are on older versions. If we could update it to the newest packages, it would be awesome as we can make use of the newer types from both these packages and get rid of any type errors for users consuming this package.
> > 
> > 
> > I just updated mongoose from v5 to v7, which is latest. I also updated mongodb from v3 to v4. However, the latest version is v5 but I was unable to get that to work because of an ongoing issue with BSON v5 found [here](https://github.com/nextauthjs/next-auth/issues/7022)
> 
> I think we resolved this by updating the Bson package. I think that should resolve the issue.

Yeah that seemed to work, thank you. I updated bson as well the other day, but it was still throwing me the same error at the time. All good now, mongodb is latest version.

devileye98 commented:
@lorensr , can we get this PR merged since I think all the comments are addressed?

NickDuBois commented:
Any hope of this getting merged?

lorensr commented:
Sorry folks, very low on time these days! Lmk if anyone's interested in helping out with maintenance. Thanks for the ping Nick. 

`0.6.0` is now released with this PR. Please comment with how it works for you!

https://github.com/GraphQLGuide/apollo-datasource-mongodb/releases/tag/0.6.0

https://www.npmjs.com/package/apollo-datasource-mongodb?activeTab=versions
